bash_completion

```diff
@@ -2614,6 +2612,17 @@ _complete_as_root()
     [[ $EUID -eq 0 || ${root_command-} ]]
 }
 
+# Complete on available commands, subject to `no_empty_cmd_completion`.
+#
+# @since 2.12
+_comp_compgen_commands()
+{
+    [[ ! ${cur-} ]] && shopt -q no_empty_cmd_completion && return
```

Comment:
If we make it the same as `_comp_compgen -- [...]` (i.e., 0 only when at least one completion is generated), we can return `1` for `no_empty_cmd_completion`.

```suggestion
    [[ ! ${cur-} ]] && shopt -q no_empty_cmd_completion && return 1
```

Reply:
SGTM in principle, although returning 1 feels a bit "wrong" in the sense that it's not an error, it's just that the completions are inhibited because of config. There probably isn't such a use case in-tree at least yet, but a pattern like `_comp_compgen_command || do_something_else_to_generate_completions` doesn't seem completely unrealistic, but we might not want to enter the "or" branch when the nonzero is due to config. In that sense, returning 0 here could be worth considering.

But I guess it could be argued the other way in favor of a nonzero. And that'd be kind of nice for consistency, too. No strong feelings on this one here.

Reply:
I have already let `_comp_split` and `_comp_compgen -- [...]` use exit status 1 when no completions are generated although it's not an error. Usage error is notified by status 2. A typical use case for `return 1` that I assume is `_comp_compgen -v command && process "${arr[@]}"`. But if there is another use case, there's no problem in returning 0 here.

I guess whether to run the "or" branch depends on the actual action in `do_something_elese_to_generate_completions`. For example, if `do_something_elese_to_generate_completions` has a small computational cost, I think it can be run even if `_comp_compgen_command` is canceled because the purpose of `no_empty_cmd_completion` is not to block the completion.